### PR TITLE
[Domain Focus] Card Logic

### PR DESCRIPTION
**Pull request by alpavanoglu**
This PR is a follow up for [PR 21241](https://github.com/wordpress-mobile/WordPress-iOS/pull/21241). It implements the logic for the card along with tracks.

Simulator Screenshot - iPhone 14 Pro - 2023-08-08 at 12 40 02

## Testing:

1. Install & Launch Jetpack
2. Enable the feature flag.
3. Navigate to MySite (either kill the app first or open a different tab and come back to My Site).
4. Verify the card is being shown.
5. Verify the track `dashboard_card_domain_transfer_shown` is fired.

### Button & Card
- Verify tapping on the button & the card opens the [correct url](https://wordpress.com/transfer-google-domains/) in webview.
- Verify the track `dashboard_card_domain_transfer_button_tapped` is fired.

### More Button
- Verify if the more button opens a menu that contains the hide functionality.
- Verify the track `dashboard_card_domain_transfer_more_menu_tapped` is fired.

### Hide
- Tap one more & hide the card.
- Verify the default hidden card track is fired for Domain Focus Card.

### WordPress App
- The card shouldn't be visible on the WordPress app.

## Regression Notes
1. Potential unintended areas of impact
The card could appear in the wrong order. 

6. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual tested.

7. What automated tests I added (or what prevented me from doing so)
It is a predominantly UI addition along with tracks. I didn't add unit tests.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by alpavanoglu** [comment]
There's some logs coming due to calls to `canOpenURL` function. This is triggered from `MigrationAppDetection.isCounterpartAppInstalled` line in `WebKitViewController` file. There's a comment saying that the logic there is a work around for an issue with universal link and can/will be removed once we get rid of universal links. I'll try to find out if we must do something about that here.

**Comment by alpavanoglu** [comment]
> There's some logs coming due to calls to `canOpenURL` function. This is triggered from `MigrationAppDetection.isCounterpartAppInstalled` line in `WebKitViewController` file. There's a comment saying that the logic there is a work around for an issue with universal link and can/will be removed once we get rid of universal links. I'll try to find out if we must do something about that here.

@dvdchr has confirmed that that "...this issue should only be reported in Debug build because the scheme isn't registered in LSApplicationQueriesSchemes in the .plist. It shouldn't cause an issue in prod".

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/DashboardGoogleDomainsCardView.swift [hunk 1]
```diff
@@ -32,6 +41,7 @@ struct DashboardGoogleDomainsCardView: View {
         HStack {
             Button {
                 print("Transfer your domains tapped")
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/DashboardGoogleDomainsCardView.swift]: Probably this print should be removed or refactored to `DDLogInfo` if you prefer to log that message.

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/DashboardGoogleDomainsCardView.swift [hunk 2]
```diff
@@ -1,13 +1,22 @@
 import SwiftUI
 
 struct DashboardGoogleDomainsCardView: View {
+    private var buttonClosure: () -> ()
+
+    init(buttonClosure: @escaping () -> ()) {
+        self.buttonClosure = buttonClosure
+    }
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/DashboardGoogleDomainsCardView.swift:8]: 💡 This is a minor suggestion. Personally, I'd prefer to replace the `Closure` suffix with `Action`. So the property name becomes `buttonAction`. This naming convention can also be found in UIKit and SwiftUI. For example, in `SwiftUI.Button` initializer, the param is named `action`.

**Comment by salimbraksa** [comment]
@alpavanoglu Code changes look good, and the test instructions work as expected. But it would be great to have unit tests for the following cases. 

1. Check if the event `dashboard_card_domain_transfer_shown`  is fired when the view appears.
2. Check if tapping the cell presents the web view and fires the `dashboard_card_domain_transfer_button_tapped` event.
3. Check if tapping the `More` button shows the `Hide this` button and fires `dashboard_card_domain_transfer_more_menu_tapped` event.

This is my opinion, but I'll let you and @hassaanelgarem decide whether unit tests are needed.

**Comment by alpavanoglu** [comment]
@salimbraksa 
> But it would be great to have unit tests for the following cases.

How do you suggest we test these cases considering the structure of the parent VC and how the Cells are created?

They are surely valid things to test. It seemed to me though they would require quite a bit of dancing around the code to test some tracks being fired. Perhaps you have a simple way around it in your mind though. In that case, we should do it.

**Comment by salimbraksa** [comment]
>  Perhaps you have a simple way around it in your mind though. In that case, we should do it.

@alpavanoglu Indeed, I've done something similar in the EU/US Compliance project. Here is how we can test the Track events.

First, we should create a type responsible only for firing Track events:

```swift
// The `WPAnalyticsEvent` enum is massive. It would be great to declare Google Domains events in a separate enum.
enum GoogleDomainsAnalytics: String {
   // ...
}

class GoogleDomainsAnalyticsTracker {
     
    private let tracker: AnalyticsEventTracking.Type

    init(tracker: AnalyticsEventTracking.Type = WPAnalytics.self) {
        self.tracker = tracker
    }

    func track(_ event: GoogleDomainsAnalyticsEvent, properties: [String: String]) {
        let event = AnalyticsEvent(name: event.rawValue, properties: properties)
        tracker.track(event)
    }
}
```
Then writing unit tests for this type becomes easy. An example of this approach is the `PrivacySettingsAnalyticsTracker` and `PrivacySettingsAnalyticsTrackerTests` classes. _( N.B. Having a protocol for the `PrivacySettingsAnalyticsTracker` was unnecessary in my opinion. We can mock that class without the protocol )._

After that, we need to write unit tests checking that this `GoogleDomainsAnalyticsTracker` instance is called in `DashboardGoogleDomainsCardCell`. But first, we should tweak `DashboardGoogleDomainsCardCell` to make it testable:

1. `GoogleDomainsAnalyticsTracker` should be injected as a dependency in `DashboardGoogleDomainsCardCell`.
2. Expose `DashboardGoogleDomainsCardCell` action handlers as callbacks:
```swift
final class DashboardGoogleDomainsCardCell: DashboardCollectionViewCell {
   
   private(set) var onViewTapped: ActionHandler?
   private(set) var onTransferButtonTapped: ActionHandler?
   private(set) var onMoreButtonTapped: ActionHandler?
   
   typealias ActionHandler = () -> Void
   
}
```
3. Implement a Spy for the `GoogleDomainsAnalyticsTracker`. See `PrivacySettingsAnalyticsTrackerSpy` for example.
4. Write unit tests checking the behavior of those action handlers. It should look like this:
```swift
final class DashboardGoogleDomainsCardCellTests: XCTestCast {
     func testViewTapped() {
            // Given
            let presentingViewController = UIViewController()
            let trackersSpy = GoogleDomainsAnalyticsTrackerSpy()
            let cell = DashboardGoogleDomainsCardCell(...)
            
            // When
            cell.onViewTapped?()

            // Then
            // Expect the correct event to be fired.
            // Expect a web view to be presented.
     }
}
```

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/DashboardGoogleDomainsCardCell.swift [hunk 3]
```diff
@@ -33,16 +36,48 @@ final class DashboardGoogleDomainsCardCell: DashboardCollectionViewCell {
...
+    }
+
+    private func presentGoogleDomainsWebView() {
+        guard let url = URL(string: Constants.transferDomainsURL) else {
+            return
+        }
+
+        let webViewController = WebViewControllerFactory.controllerAuthenticatedWithDefaultAccount(
+            url: url,
+            source: "domain_focus_card"
```
Comment (hassaanelgarem) [inline WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/DashboardGoogleDomainsCardCell.swift:85]: [Nitpick]
```suggestion
            source: "domain_transfer_card"
```

**Review by hassaanelgarem — APPROVED** [review]
Works as described 🚀

**Comment by hassaanelgarem** [comment]
> I'll let you and @hassaanelgarem decide whether unit tests are needed.

Adding unit tests is always a good idea, and the strategy Salim outlined makes sense to me.

@alpavanoglu It's your call whether you see the extra effort is worth it 👍

**Comment by hassaanelgarem** [comment]
@alpavanoglu I accidentally merged this while reviewing 🙈 
Sorry for this!

**Pull request merged**